src/sinks/greptimedb/service.rs

```diff
@@ -0,0 +1,188 @@
+    }
+
+    fn bytes_sent(&self) -> Option<usize> {
+        Some(self.metadata.request_wire_size())
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct GreptimeDBService {
+    /// the client that connects to greptimedb
+    client: Arc<Database>,
+}
+
+impl GreptimeDBService {
+    pub fn try_new(config: &GreptimeDBConfig) -> crate::Result<Self> {
+        let grpc_client = if let Some(tls_config) = &config.tls {
+            let channel_config = ChannelConfig {
+                client_tls: Self::try_from_tls_config(tls_config)?,
+                ..Default::default()
+            };
+            Client::with_manager_and_urls(
+                ChannelManager::with_tls_config(channel_config).map_err(Box::new)?,
+                vec![&config.endpoint],
+            )
+        } else {
+            Client::with_urls(vec![&config.endpoint])
+        };
+
+        let mut client = Database::new_with_dbname(&config.dbname, grpc_client);
+
+        if let (Some(username), Some(password)) = (&config.username, &config.password) {
+            client.set_auth(AuthScheme::Basic(Basic {
+                username: username.to_owned(),
+                password: password.clone().into(),
+            }))
+        };
+
+        Ok(GreptimeDBService {
+            client: Arc::new(client),
+        })
+    }
+
+    fn try_from_tls_config(tls_config: &TlsConfig) -> crate::Result<Option<ClientTlsOption>> {
+        if let Some(ca_path) = tls_config.ca_file.as_ref() {
+            let cert_path = tls_config
+                .crt_file
+                .as_ref()
+                .ok_or(GreptimeDBConfigError::TlsMissingCert)?;
+            let key_path = tls_config
+                .key_file
+                .as_ref()
+                .ok_or(GreptimeDBConfigError::TlsMissingKey)?;
+
+            if tls_config.key_pass.is_some() {
+                warn!(
+                    message = "TLS key file with password is not supported by greptimedb client at the moment."
+                );
+            }
+
+            Ok(Some(ClientTlsOption {
+                server_ca_cert_path: ca_path.clone(),
+                client_key_path: key_path.clone(),
+                client_cert_path: cert_path.clone(),
+            }))
+        } else {
+            Ok(None)
+        }
+    }
+}
+
+impl Service<GreptimeDBRequest> for GreptimeDBService {
+    type Response = GreptimeDBBatchOutput;
+    type Error = GreptimeError;
+    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;
+
+    fn poll_ready(&mut self, _cx: &mut std::task::Context) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
+    }
+
+    // Convert vector metrics into GreptimeDB format and send them in batch
+    fn call(&mut self, req: GreptimeDBRequest) -> Self::Future {
+        let client = Arc::clone(&self.client);
+
+        Box::pin(async move {
+            let metadata = req.get_metadata().clone();
+            let result = client.insert(req.items).await?;
```

Comment:
Small nit, you should be able to avoid this clone by accessing the field directly.

```suggestion
            let metadata = req.metadata;
            let result = client.insert(req.items).await?;
```